libioc/Provisioning/puppet.py

```diff
@@ -0,0 +1,195 @@
+        return str(self._url)
+
+    @property.setter
+    def url(self) -> str:
+        """Set the Puppet Control-Repo URL."""
+        return str(self._url)
+
+    @property
+    def name(self, value: str) -> str:
+        """Return the unique name for this Puppet Control-Repo URL."""
+        return self._name
+
+    @property.setter
+    def name(self, value: str) -> None:
+        """Set a unique name for this Puppet Control-Repo URL."""
+        self._name = value
+
+    @property
+    def pkgs(self, value: str) -> typing.List[str]:
+        """Return list of packages required for this Provisioning method."""
+        return self._pkgs
+
+    #@property.setter
+    #def pkgs(self, value: str) -> None:
+    #    """Set (list) of additional packagess required for this Puppet Control-Repo URL."""
+    #    self._pkgs += value
+
+
+def provision(
+    self: 'libioc.Provisioning.Prototype',
+    event_scope: typing.Optional['libioc.events.Scope']=None
+) -> typing.Generator['libioc.events.IocEvent', None, None]:
+    """
+    Provision the jail with Puppet apply using the supplied control-repo.
+
+    The repo can either be a filesystem path, or a http[s]/git URL.
+    If the repo is a filesystem path, it will be mounted appropriately.
+    If the repo is a URL, it will be setup with `r10k`.
+
+        ioc set \
+            provisioning.method=puppet \
+            provisioning.source=http://example.com/my/puppet-env \
+            myjail
+
+    """
+    events = libioc.events
+    jailProvisioningEvent = events.JailProvisioning(
+        jail=self.jail,
+        event_scope=event_scope
+    )
+    yield jailProvisioningEvent.begin()
+    _scope = jailProvisioningEvent.scope
+    jailProvisioningAssetDownloadEvent = events.JailProvisioningAssetDownload(
+        jail=self.jail,
+        event_scope=_scope
+    )
+
+    # download / mount provisioning assets
+    try:
+        yield jailProvisioningAssetDownloadEvent.begin()
+        pluginDefinition = ControlRepoDefinition(
+            self.source,
+            logger=self.jail.logger
+        )
+        yield jailProvisioningAssetDownloadEvent.end()
+    except Exception as e:
+        yield jailProvisioningAssetDownloadEvent.fail(e)
+        raise e
+
+    if not (self.source.startswith('file://') or self.source.startswith('/')):
+        # clone control repo
+        controlrepo_dataset_name = f"{self.jail.dataset.name}/puppet"
+        controlrepo_dataset = self.zfs.get_empty_dataset(
+            control_repo_dataset_name
+        )
+
+        git.Repo.clone_from(
+            ControlRepoDefinition["url"],
+            plugin_dataset.mountpoint
+        )
+
+    self.jail.fstab.new_line(
+        source=plugin_dataset.mountpoint,
+        destination="/.puppet",
+        options="ro",
+        auto_create_destination=True,
+        replace=True
+    )
+    self.jail.fstab.save()
+
+    if "pkgs" in controlRepoDefinition.keys():
+        pkg_packages = list(controlRepoDefinition["pkgs"])
```

Comment:
pkgs is a property, not a a dict item. It needs to be accessed as `controlRepoDefinition.pkg`.